File: src-tauri/src/domain/sort_modes.rs
//! Usage: Sort mode persistence and provider ordering configuration helpers.

use crate::db;
use crate::shared::error::db_err;
use crate::shared::time::now_unix_seconds;
use rusqlite::{params, params_from_iter, Connection, OptionalExtension};
use serde::Serialize;
use std::collections::{HashMap, HashSet};

const MAX_SORT_MODE_NAME_CHARS: usize = 32;
const MAX_SORT_MODE_PROVIDER_IDS: usize = 512;

#[derive(Debug, Clone, Serialize, specta::Type)]
pub struct SortModeSummary {
    pub id: i64,
    pub name: String,
    pub created_at: i64,
    pub updated_at: i64,
}

#[derive(Debug, Clone, Serialize, specta::Type)]
pub struct SortModeActiveRow {
    pub cli_key: String,
    pub mode_id: Option<i64>,
    pub updated_at: i64,
}

#[derive(Debug, Clone, Serialize, specta::Type)]
pub struct SortModeProviderRow {
    pub provider_id: i64,
    pub enabled: bool,
}

fn enabled_to_int(enabled: bool) -> i64 {
    if enabled {
        1
    } else {
        0
    }
}

fn enabled_from_int(value: i64) -> bool {
    value != 0
}

fn validate_cli_key(cli_key: &str) -> crate::shared::error::AppResult<()> {
    crate::shared::cli_key::validate_cli_key(cli_key)
}

fn validate_mode_name(name: &str) -> crate::shared::error::AppResult<String> {
    let name = name.trim();
    if name.is_empty() {
        return Err("SEC_INVALID_INPUT: mode name is required".into());
    }

    if name.chars().nth(MAX_SORT_MODE_NAME_CHARS).is_some() {
        return Err(format!(
            "SEC_INVALID_INPUT: mode name is too long (max {MAX_SORT_MODE_NAME_CHARS} chars)"
        )
        .into());
    }

    let lowered = name.to_ascii_lowercase();
    if lowered == "default" || name == "默认" {
        return Err("SEC_INVALID_INPUT: mode name is reserved".into());
    }

    Ok(name.to_string())
}

fn row_to_mode_summary(row: &rusqlite::Row<'_>) -> Result<SortModeSummary, rusqlite::Error> {
    Ok(SortModeSummary {
        id: row.get("id")?,
        name: row.get("name")?,
        created_at: row.get("created_at")?,
        updated_at: row.get("updated_at")?,
    })
}

fn ensure_mode_exists(conn: &Connection, mode_id: i64) -> crate::shared::error::AppResult<()> {
    if mode_id <= 0 {
        return Err("SEC_INVALID_INPUT: invalid mode_id".into());
    }

    let exists: Option<i64> = conn
        .query_row(
            "SELECT id FROM sort_modes WHERE id = ?1",
            params![mode_id],
            |row| row.get(0),
        )
        .optional()
        .map_err(|e| db_err!("failed to query sort_mode: {e}"))?;

    if exists.is_none() {
        return Err("DB_NOT_FOUND: sort_mode not found".into());
    }

    Ok(())
}

fn read_active_row(
    conn: &Connection,
    cli_key: &str,
) -> crate::shared::error::AppResult<SortModeActiveRow> {
    conn.query_row(
        r#"
SELECT
  cli_key,
  mode_id,
  updated_at
FROM sort_mode_active
WHERE cli_key = ?1
"#,
        params![cli_key],
        |row| {
            Ok(SortModeActiveRow {
                cli_key: row.get("cli_key")?,
                mode_id: row.get("mode_id")?,
                updated_at: row.get("updated_at")?,
            })
        },
    )
    .optional()
    .map_err(|e| db_err!("failed to query sort_mode_active: {e}"))?
    .ok_or_else(|| "DB_NOT_FOUND: sort_mode_active not found".into())
}

pub fn list_modes(db: &db::Db) -> crate::shared::error::AppResult<Vec<SortModeSummary>> {
    let conn = db.open_connection()?;
    let mut stmt = conn
        .prepare_cached(
            r#"
SELECT
  id,
  name,
  created_at,
  updated_at
FROM sort_modes
ORDER BY id ASC
"#,
        )
        .map_err(|e| db_err!("failed to prepare sort_modes query: {e}"))?;

    let rows = stmt
        .query_map([], row_to_mode_summary)
        .map_err(|e| db_err!("failed to list sort_modes: {e}"))?;

    let mut items = Vec::new();
    for row in rows {
        items.push(row.map_err(|e| db_err!("failed to read sort_mode row: {e}"))?);
    }
    Ok(items)
}

pub fn create_mode(db: &db::Db, name: &str) -> crate::shared::error::AppResult<SortModeSummary> {
    let name = validate_mode_name(name)?;
    let conn = db.open_connection()?;
    let now = now_unix_seconds();

    conn.execute(
        r#"
INSERT INTO sort_modes(
  name,
  created_at,
  updated_at
) VALUES (?1, ?2, ?3)
"#,
        params![name, now, now],
    )
    .map_err(|e| match e {
        rusqlite::Error::SqliteFailure(err, _)
            if err.code == rusqlite::ErrorCode::ConstraintViolation =>
        {
            crate::shared::error::AppError::new(
                "DB_CONSTRAINT",
                format!("sort_mode already exists: name={name}"),
            )
        }
        other => db_err!("failed to insert sort_mode: {other}"),
    })?;

    let id = conn.last_insert_rowid();
    conn.query_row(
        r#"
SELECT
  id,
  name,
  created_at,
  updated_at
FROM sort_modes
WHERE id = ?1
"#,
        params![id],
        row_to_mode_summary,
    )
    .map_err(|e| db_err!("failed to query inserted sort_mode: {e}"))
}

pub fn rename_mode(
    db: &db::Db,
    mode_id: i64,
    name: &str,
) -> crate::shared::error::AppResult<SortModeSummary> {
    let name = validate_mode_name(name)?;
    let conn = db.open_connection()?;
    ensure_mode_exists(&conn, mode_id)?;
    let now = now_unix_seconds();

    conn.execute(
        "UPDATE sort_modes SET name = ?1, updated_at = ?2 WHERE id = ?3",
        params![name, now, mode_id],
    )
    .map_err(|e| match e {
        rusqlite::Error::SqliteFailure(err, _)
            if err.code == rusqlite::ErrorCode::ConstraintViolation =>
        {
            crate::shared::error::AppError::new(
                "DB_CONSTRAINT",
                format!("sort_mode already exists: name={name}"),
            )
        }
        other => db_err!("failed to update sort_mode: {other}"),
    })?;

    conn.query_row(
        r#"
SELECT
  id,
  name,
  created_at,
  updated_at
FROM sort_modes
WHERE id = ?1
"#,
        params![mode_id],
        row_to_mode_summary,
    )
    .map_err(|e| db_err!("failed to query sort_mode: {e}"))
}

pub fn delete_mode_with_affected_cli_keys(
    db: &db::Db,
    mode_id: i64,
) -> crate::shared::error::AppResult<Vec<String>> {
    let conn = db.open_connection()?;
    ensure_mode_exists(&conn, mode_id)?;

    let mut stmt = conn
        .prepare_cached(
            r#"
SELECT cli_key
FROM sort_mode_active
WHERE mode_id = ?1
ORDER BY cli_key ASC
"#,
        )
        .map_err(|e| db_err!("failed to prepare active sort_mode query: {e}"))?;
    let rows = stmt
        .query_map(params![mode_id], |row| row.get::<_, String>(0))
        .map_err(|e| db_err!("failed to query active sort_mode cli keys: {e}"))?;

    let mut affected_cli_keys = Vec::new();
    for row in rows {
        affected_cli_keys
            .push(row.map_err(|e| db_err!("failed to read active sort_mode cli key: {e}"))?);
    }
    drop(stmt);

    let changed = conn
        .execute("DELETE FROM sort_modes WHERE id = ?1", params![mode_id])
        .map_err(|e| db_err!("failed to delete sort_mode: {e}"))?;
    if changed == 0 {
        return Err("DB_NOT_FOUND: sort_mode not found".to_string().into());
    }
    Ok(affected_cli_keys)
}

pub fn delete_mode(db: &db::Db, mode_id: i64) -> crate::shared::error::AppResult<()> {
    delete_mode_with_affected_cli_keys(db, mode_id)?;
    Ok(())
}

pub fn list_active(db: &db::Db) -> crate::shared::error::AppResult<Vec<SortModeActiveRow>> {
    let conn = db.open_connection()?;
    let mut stmt = conn
        .prepare_cached(
            r#"
SELECT
  cli_key,
  mode_id,
  updated_at
FROM sort_mode_active
ORDER BY cli_key ASC
"#,
        )
        .map_err(|e| db_err!("failed to prepare sort_mode_active query: {e}"))?;

    let rows = stmt
        .query_map([], |row| {
            Ok(SortModeActiveRow {
                cli_key: row.get("cli_key")?,
                mode_id: row.get("mode_id")?,
                updated_at: row.get("updated_at")?,
            })
        })
        .map_err(|e| db_err!("failed to list sort_mode_active: {e}"))?;

    let mut items = Vec::new();
    for row in rows {
        items.push(row.map_err(|e| db_err!("failed to read sort_mode_active row: {e}"))?);
    }
    Ok(items)
}

pub fn set_active(
    db: &db::Db,
    cli_key: &str,
    mode_id: Option<i64>,
) -> crate::shared::error::AppResult<SortModeActiveRow> {
    let cli_key = cli_key.trim();
    validate_cli_key(cli_key)?;

    let conn = db.open_connection()?;
    if let Some(mode_id) = mode_id {
        ensure_mode_exists(&conn, mode_id)?;
    }
    let now = now_unix_seconds();

    conn.execute(
        r#"
INSERT INTO sort_mode_active(
  cli_key,
  mode_id,
  updated_at
) VALUES (?1, ?2, ?3)
ON CONFLICT(cli_key) DO UPDATE SET
  mode_id = excluded.mode_id,
  updated_at = excluded.updated_at
"#,
        params![cli_key, mode_id, now],
    )
    .map_err(|e| db_err!("failed to upsert sort_mode_active: {e}"))?;

    read_active_row(&conn, cli_key)
}

pub fn list_mode_providers(
    db: &db::Db,
    mode_id: i64,
    cli_key: &str,
) -> crate::shared::error::AppResult<Vec<SortModeProviderRow>> {
    let cli_key = cli_key.trim();
    validate_cli_key(cli_key)?;
    let conn = db.open_connection()?;
    ensure_mode_exists(&conn, mode_id)?;

    let mut stmt = conn
        .prepare_cached(
            r#"
SELECT
  provider_id,
  enabled
FROM sort_mode_providers
WHERE mode_id = ?1
  AND cli_key = ?2
ORDER BY sort_order ASC
"#,
        )
        .map_err(|e| db_err!("failed to prepare sort_mode_providers query: {e}"))?;

    let rows = stmt
        .query_map(params![mode_id, cli_key], |row| {
            let provider_id: i64 = row.get(0)?;
            let enabled_raw: i64 = row.get(1)?;
            Ok(SortModeProviderRow {
                provider_id,
                enabled: enabled_from_int(enabled_raw),
            })
        })
        .map_err(|e| db_err!("failed to list sort_mode_providers: {e}"))?;

    let mut items = Vec::new();
    for row in rows {
        items.push(row.map_err(|e| db_err!("failed to read sort_mode_provider row: {e}"))?);
    }
    Ok(items)
}

fn ensure_providers_belong_to_cli(
    conn: &Connection,
    cli_key: &str,
    provider_ids: &[i64],
) -> crate::shared::error::AppResult<()> {
    if provider_ids.is_empty() {
        return Ok(());
    }

    if provider_ids.len() > MAX_SORT_MODE_PROVIDER_IDS {
        return Err(format!(
            "SEC_INVALID_INPUT: ordered_provider_ids must contain at most {MAX_SORT_MODE_PROVIDER_IDS} entries"
        )
        .into());
    }

    let mut unique_ids = HashSet::new();
    for id in provider_ids {
        if *id <= 0 {
            return Err(format!("SEC_INVALID_INPUT: invalid provider_id={id}").into());
        }
        if !unique_ids.insert(*id) {
            return Err(format!("SEC_INVALID_INPUT: duplicate provider_id={id}").into());
        }
    }

    let placeholders = db::sql_placeholders(unique_ids.len());
    let sql = format!("SELECT id FROM providers WHERE cli_key = ?1 AND id IN ({placeholders})");

    let mut stmt = conn
        .prepare(&sql)
        .map_err(|e| db_err!("failed to prepare provider validation query: {e}"))?;

    let mut params_vec: Vec<rusqlite::types::Value> = Vec::with_capacity(unique_ids.len() + 1);
    params_vec.push(rusqlite::types::Value::from(cli_key.to_string()));
    params_vec.extend(unique_ids.iter().map(|id| (*id).into()));

    let rows = stmt
        .query_map(params_from_iter(params_vec), |row| row.get::<_, i64>(0))
        .map_err(|e| db_err!("failed to query provider validation: {e}"))?;

    let mut found = HashSet::new();
    for row in rows {
        found.insert(row.map_err(|e| db_err!("failed to read provider id: {e}"))?);
    }

    if found.len() != unique_ids.len() {
        let missing: Vec<i64> = unique_ids.difference(&found).copied().collect();
        return Err(format!(
            "SEC_INVALID_INPUT: provider_id does not belong to cli_key={cli_key}: {missing:?}"
        )
        .into());
    }

    Ok(())
}

pub fn set_mode_providers_order(
    db: &db::Db,
    mode_id: i64,
    cli_key: &str,
    ordered_provider_ids: Vec<i64>,
) -> crate::shared::error::AppResult<Vec<SortModeProviderRow>> {
    let cli_key = cli_key.trim();
    validate_cli_key(cli_key)?;

    let mut conn = db.open_connection()?;
    ensure_mode_exists(&conn, mode_id)?;
    ensure_providers_belong_to_cli(&conn, cli_key, &ordered_provider_ids)?;

    let tx = conn
        .transaction()
        .map_err(|e| db_err!("failed to start transaction: {e}"))?;

    let mut existing_enabled: HashMap<i64, bool> = HashMap::new();
    {
        let mut stmt = tx
            .prepare_cached(
                r#"
SELECT
  provider_id,
  enabled
FROM sort_mode_providers
WHERE mode_id = ?1
  AND cli_key = ?2
"#,
            )
            .map_err(|e| db_err!("failed to prepare sort_mode_providers query: {e}"))?;
        let rows = stmt
            .query_map(params![mode_id, cli_key], |row| {
                let provider_id: i64 = row.get(0)?;
                let enabled_raw: i64 = row.get(1)?;
                Ok((provider_id, enabled_from_int(enabled_raw)))
            })
            .map_err(|e| db_err!("failed to list sort_mode_providers: {e}"))?;
        for row in rows {
            let (provider_id, enabled) =
                row.map_err(|e| db_err!("failed to read sort_mode_provider row: {e}"))?;
            existing_enabled.insert(provider_id, enabled);
        }
    }

    tx.execute(
        "DELETE FROM sort_mode_providers WHERE mode_id = ?1 AND cli_key = ?2",
        params![mode_id, cli_key],
    )
    .map_err(|e| db_err!("failed to clear sort_mode_providers: {e}"))?;

    let now = now_unix_seconds();
    for (idx, provider_id) in ordered_provider_ids.iter().enumerate() {
        let enabled = existing_enabled.get(provider_id).copied().unwrap_or(true);
        tx.execute(
            r#"
INSERT INTO sort_mode_providers(
  mode_id,
  cli_key,
  provider_id,
  sort_order,
  enabled,
  created_at,
  updated_at
) VALUES (?1, ?2, ?3, ?4, ?5, ?6, ?7)
"#,
            params![
                mode_id,
                cli_key,
                provider_id,
                idx as i64,
                enabled_to_int(enabled),
                now,
                now
            ],
        )
        .map_err(|e| db_err!("failed to insert sort_mode_provider: {e}"))?;
    }

    tx.commit()
        .map_err(|e| db_err!("failed to commit transaction: {e}"))?;

    list_mode_providers(db, mode_id, cli_key)
}

pub fn set_mode_provider_enabled(
    db: &db::Db,
    mode_id: i64,
    cli_key: &str,
    provider_id: i64,
    enabled: bool,
) -> crate::shared::error::AppResult<SortModeProviderRow> {
    let cli_key = cli_key.trim();
    validate_cli_key(cli_key)?;
    if provider_id <= 0 {
        return Err("SEC_INVALID_INPUT: invalid provider_id".into());
    }

    let conn = db.open_connection()?;
    ensure_mode_exists(&conn, mode_id)?;
    ensure_providers_belong_to_cli(&conn, cli_key, &[provider_id])?;

    let now = now_unix_seconds();
    let changed = conn
        .execute(
            r#"
UPDATE sort_mode_providers
SET enabled = ?1, updated_at = ?2
WHERE mode_id = ?3
  AND cli_key = ?4
  AND provider_id = ?5
"#,
            params![enabled_to_int(enabled), now, mode_id, cli_key, provider_id],
        )
        .map_err(|e| db_err!("failed to update sort_mode_provider: {e}"))?;
    if changed == 0 {
        return Err("DB_NOT_FOUND: sort_mode_provider not found".into());
    }

    conn.query_row(
        r#"
SELECT
  provider_id,
  enabled
FROM sort_mode_providers
WHERE mode_id = ?1
  AND cli_key = ?2
  AND provider_id = ?3
"#,
        params![mode_id, cli_key, provider_id],
        |row| {
            let provider_id: i64 = row.get(0)?;
            let enabled_raw: i64 = row.get(1)?;
            Ok(SortModeProviderRow {
                provider_id,
                enabled: enabled_from_int(enabled_raw),
            })
        },
    )
    .map_err(|e| db_err!("failed to read sort_mode_provider: {e}"))
}

#[cfg(test)]
mod tests {
    use super::*;
    use rusqlite::params;

    fn active_mode_id(db: &db::Db, cli_key: &str) -> Option<i64> {
        let conn = db.open_connection().expect("open db");
        conn.query_row(
            "SELECT mode_id FROM sort_mode_active WHERE cli_key = ?1",
            params![cli_key],
            |row| row.get::<_, Option<i64>>(0),
        )
        .expect("read active mode")
    }

    #[test]
    fn delete_mode_returns_active_cli_keys_before_fk_nulls_active_rows() {
        let dir = tempfile::tempdir().expect("tempdir");
        let db_path = dir.path().join("sort_mode_delete_active_cli_keys.db");
        let db = db::init_for_tests(&db_path).expect("init db");

        let deleted_mode = create_mode(&db, "Review Mode").expect("create deleted mode");
        let other_mode = create_mode(&db, "Other Mode").expect("create other mode");
        set_active(&db, "claude", Some(deleted_mode.id)).expect("activate claude");
        set_active(&db, "codex", Some(deleted_mode.id)).expect("activate codex");
        set_active(&db, "gemini", Some(other_mode.id)).expect("activate gemini");

        let affected_cli_keys =
            delete_mode_with_affected_cli_keys(&db, deleted_mode.id).expect("delete mode");

        assert_eq!(
            affected_cli_keys,
            vec!["claude".to_string(), "codex".to_string()]
        );
        assert_eq!(active_mode_id(&db, "claude"), None);
        assert_eq!(active_mode_id(&db, "codex"), None);
        assert_eq!(active_mode_id(&db, "gemini"), Some(other_mode.id));
    }
}

File: src/services/providers/sortModes.ts
import { commands } from "../../generated/bindings";
import { FeValidationError } from "../../utils/errors";
import { invokeGeneratedIpc, type GeneratedCommandResult } from "../generatedIpc";
import { validateProviderCliKey, type CliKey } from "./providers";

export const MAX_SORT_MODE_NAME_CHARS = 32;
export const MAX_SORT_MODE_PROVIDER_IDS = 512;

export type SortModeSummary = {
  id: number;
  name: string;
  created_at: number;
  updated_at: number;
};

export type SortModeActiveRow = {
  cli_key: CliKey;
  mode_id: number | null;
  updated_at: number;
};

export type SortModeProviderRow = {
  provider_id: number;
  enabled: boolean;
};

function normalizeSortModeName(name: string) {
  const trimmed = name.trim();
  if (!trimmed) {
    throw new Error("SEC_INVALID_INPUT: mode name is required");
  }
  if ([...trimmed].length > MAX_SORT_MODE_NAME_CHARS) {
    throw new Error(
      `SEC_INVALID_INPUT: mode name is too long (max ${MAX_SORT_MODE_NAME_CHARS} chars)`
    );
  }
  if (trimmed.toLowerCase() === "default" || trimmed === "默认") {
    throw new Error("SEC_INVALID_INPUT: mode name is reserved");
  }
  return trimmed;
}

function validatePositiveId(field: string, value: number) {
  if (!Number.isSafeInteger(value) || value <= 0) {
    throw new FeValidationError(`SEC_INVALID_INPUT: invalid ${field}=${value}`);
  }
}

export function validateSortModeId(modeId: number): number {
  validatePositiveId("modeId", modeId);
  return modeId;
}

function validateOrderedProviderIds(orderedProviderIds: number[]) {
  if (orderedProviderIds.length > MAX_SORT_MODE_PROVIDER_IDS) {
    throw new Error(
      `SEC_INVALID_INPUT: orderedProviderIds must contain at most ${MAX_SORT_MODE_PROVIDER_IDS} entries`
    );
  }

  const seen = new Set<number>();
  for (const providerId of orderedProviderIds) {
    validatePositiveId("providerId", providerId);
    if (seen.has(providerId)) {
      throw new Error(`SEC_INVALID_INPUT: duplicate providerId=${providerId}`);
    }
    seen.add(providerId);
  }
}

export async function sortModesList() {
  return invokeGeneratedIpc<SortModeSummary[]>({
    title: "读取排序模板失败",
    cmd: "sort_modes_list",
    invoke: () => commands.sortModesList() as Promise<GeneratedCommandResult<SortModeSummary[]>>,
  });
}

export async function sortModeCreate(input: { name: string }) {
  const name = normalizeSortModeName(input.name);

  return invokeGeneratedIpc<SortModeSummary>({
    title: "创建排序模板失败",
    cmd: "sort_mode_create",
    args: { name },
    invoke: () => commands.sortModeCreate(name) as Promise<GeneratedCommandResult<SortModeSummary>>,
  });
}

export async function sortModeRename(input: { mode_id: number; name: string }) {
  const modeId = validateSortModeId(input.mode_id);
  const name = normalizeSortModeName(input.name);

  return invokeGeneratedIpc<SortModeSummary>({
    title: "重命名排序模板失败",
    cmd: "sort_mode_rename",
    args: { modeId, name },
    invoke: () =>
      commands.sortModeRename(modeId, name) as Promise<GeneratedCommandResult<SortModeSummary>>,
  });
}

export async function sortModeDelete(input: { mode_id: number }) {
  const modeId = validateSortModeId(input.mode_id);

  return invokeGeneratedIpc<boolean>({
    title: "删除排序模板失败",
    cmd: "sort_mode_delete",
    args: { modeId },
    invoke: () => commands.sortModeDelete(modeId) as Promise<GeneratedCommandResult<boolean>>,
  });
}

export async function sortModeActiveList() {
  return invokeGeneratedIpc<SortModeActiveRow[]>({
    title: "读取激活排序模板失败",
    cmd: "sort_mode_active_list",
    invoke: () =>
      commands.sortModeActiveList() as Promise<GeneratedCommandResult<SortModeActiveRow[]>>,
  });
}

export async function sortModeActiveSet(input: { cli_key: CliKey; mode_id: number | null }) {
  const cliKey = validateProviderCliKey(input.cli_key);
  const modeId = input.mode_id == null ? null : validateSortModeId(input.mode_id);

  return invokeGeneratedIpc<SortModeActiveRow>({
    title: "设置激活排序模板失败",
    cmd: "sort_mode_active_set",
    args: { cliKey, modeId },
    invoke: () =>
      commands.sortModeActiveSet(cliKey, modeId) as Promise<
        GeneratedCommandResult<SortModeActiveRow>
      >,
  });
}

export async function sortModeProvidersList(input: { mode_id: number; cli_key: CliKey }) {
  const cliKey = validateProviderCliKey(input.cli_key);
  const modeId = validateSortModeId(input.mode_id);

  return invokeGeneratedIpc<SortModeProviderRow[]>({
    title: "读取排序模板供应商失败",
    cmd: "sort_mode_providers_list",
    args: { modeId, cliKey },
    invoke: () =>
      commands.sortModeProvidersList(modeId, cliKey) as Promise<
        GeneratedCommandResult<SortModeProviderRow[]>
      >,
  });
}

export async function sortModeProvidersSetOrder(input: {
  mode_id: number;
  cli_key: CliKey;
  ordered_provider_ids: number[];
}) {
  const cliKey = validateProviderCliKey(input.cli_key);
  const modeId = validateSortModeId(input.mode_id);
  validateOrderedProviderIds(input.ordered_provider_ids);

  return invokeGeneratedIpc<SortModeProviderRow[]>({
    title: "更新排序模板供应商顺序失败",
    cmd: "sort_mode_providers_set_order",
    args: {
      modeId,
      cliKey,
      orderedProviderIds: input.ordered_provider_ids,
    },
    invoke: () =>
      commands.sortModeProvidersSetOrder(modeId, cliKey, input.ordered_provider_ids) as Promise<
        GeneratedCommandResult<SortModeProviderRow[]>
      >,
  });
}

export async function sortModeProviderSetEnabled(input: {
  mode_id: number;
  cli_key: CliKey;
  provider_id: number;
  enabled: boolean;
}) {
  const cliKey = validateProviderCliKey(input.cli_key);
  const modeId = validateSortModeId(input.mode_id);
  validatePositiveId("providerId", input.provider_id);

  return invokeGeneratedIpc<SortModeProviderRow>({
    title: "更新排序模板供应商启用状态失败",
    cmd: "sort_mode_provider_set_enabled",
    args: {
      modeId,
      cliKey,
      providerId: input.provider_id,
      enabled: input.enabled,
    },
    invoke: () =>
      commands.sortModeProviderSetEnabled(
        modeId,
        cliKey,
        input.provider_id,
        input.enabled
      ) as Promise<GeneratedCommandResult<SortModeProviderRow>>,
  });
}
